File: scripts/collect_double_pendulum_data.js
#!/usr/bin/env node

/**
 * Double Pendulum Data Collection Script
 * Collects training data by running the double pendulum simulator with various initial conditions
 * Target: 20,000 samples for AI training
 */

const fs = require('fs');
const path = require('path');
const { JSDOM } = require('jsdom');

// Setup DOM environment for running browser-based simulator
const dom = new JSDOM('<!DOCTYPE html><html><body></body></html>');
global.window = dom.window;
global.document = dom.window.document;
global.console = console;

// Load the simulator
const simulatorPath = path.join(__dirname, '../double-pendulum/simulator.js');
const simulatorCode = fs.readFileSync(simulatorPath, 'utf8');
eval(simulatorCode);

const DATA_DIR = path.join(__dirname, '../ai_data/double-pendulum');
const SAMPLES_PER_TRAJECTORY = 200;  // Number of timesteps per trajectory
const TIMESTEP = 0.01;

// Ensure data directory exists
if (!fs.existsSync(DATA_DIR)) {
    fs.mkdirSync(DATA_DIR, { recursive: true });
}

console.log('='.repeat(70));
console.log('Double Pendulum AI Training Data Collection');
console.log('='.repeat(70));
console.log(`Target: 20,000 samples`);
console.log(`Data directory: ${DATA_DIR}`);
console.log('');

/**
 * Generate random initial conditions for double pendulum
 */
function generateRandomIC() {
    // Vary angles from -π to π
    const theta1 = (Math.random() - 0.5) * 2 * Math.PI;
    const theta2 = (Math.random() - 0.5) * 2 * Math.PI;
    
    // Vary angular velocities from -5 to 5 rad/s
    const omega1 = (Math.random() - 0.5) * 10;
    const omega2 = (Math.random() - 0.5) * 10;
    
    // Vary physical parameters slightly
    const l1 = 0.8 + Math.random() * 0.4;  // 0.8 to 1.2
    const l2 = 0.8 + Math.random() * 0.4;  // 0.8 to 1.2
    const m1 = 0.5 + Math.random() * 1.0;  // 0.5 to 1.5
    const m2 = 0.5 + Math.random() * 1.0;  // 0.5 to 1.5
    
    return { theta1, omega1, theta2, omega2, l1, l2, m1, m2 };
}

/**
 * Classify pendulum behavior based on motion characteristics
 */
function classifyBehavior(trajectory, energy_drift) {
    // Calculate motion statistics
    let totalAngleChange1 = 0;
    let totalAngleChange2 = 0;
    let periodicScore = 0;
    
    for (let i = 1; i < trajectory.length; i++) {
        const angleChange1 = Math.abs(trajectory[i].theta1 - trajectory[i-1].theta1);
        const angleChange2 = Math.abs(trajectory[i].theta2 - trajectory[i-1].theta2);
        
        totalAngleChange1 += angleChange1;
        totalAngleChange2 += angleChange2;
    }
    
    const avgAngleChange1 = totalAngleChange1 / trajectory.length;
    const avgAngleChange2 = totalAngleChange2 / trajectory.length;
    
    // Check for periodic motion (looking for repeating patterns)
    const omega1_values = trajectory.map(s => s.omega1);
    const omega2_values = trajectory.map(s => s.omega2);
    
    // Simple autocorrelation check for periodicity
    const mid = Math.floor(trajectory.length / 2);
    let correlation = 0;
    for (let i = 0; i < mid; i++) {
        const diff1 = Math.abs(omega1_values[i] - omega1_values[i + mid]);
        const diff2 = Math.abs(omega2_values[i] - omega2_values[i + mid]);
        correlation += (diff1 + diff2);
    }
    correlation /= mid;
    
    // Classification logic
    if (correlation < 1.0 && avgAngleChange1 < 0.5 && avgAngleChange2 < 0.5) {
        return 'periodic';
    } else if (totalAngleChange1 > 20 || totalAngleChange2 > 20) {
        // High degree of chaotic motion
        return 'chaotic';
    } else if (correlation < 2.0) {
        // Shows some periodicity but not pure periodic
        return 'resonant';
    } else {
        return 'chaotic';
    }
}

/**
 * Calculate Lyapunov exponent approximation (measure of chaos)
 */
function calculateLyapunovApproximation(trajectory) {
    let sum = 0;
    let count = 0;
    
    for (let i = 10; i < trajectory.length; i++) {
        const d1 = Math.abs(trajectory[i].theta1 - trajectory[i-10].theta1);
        const d2 = Math.abs(trajectory[i].theta2 - trajectory[i-10].theta2);
        const distance = Math.sqrt(d1*d1 + d2*d2);
        
        if (distance > 1e-10) {
            sum += Math.log(distance);
            count++;
        }
    }
    
    return count > 0 ? sum / count : 0;
}

/**
 * Run a single simulation trajectory
 */
function runSimulation(ic, trajectoryId) {
    const simulator = new DoublePendulumSimulator();
    
    // Set parameters
    simulator.l1 = ic.l1;
    simulator.l2 = ic.l2;
    simulator.m1 = ic.m1;
    simulator.m2 = ic.m2;
    simulator.dt = TIMESTEP;
    
    // Set initial conditions
    simulator.pendulums[0].state = [ic.theta1, ic.omega1, ic.theta2, ic.omega2];
    simulator.pendulums[0].l1 = ic.l1;
    simulator.pendulums[0].l2 = ic.l2;
    simulator.pendulums[0].m1 = ic.m1;
    simulator.pendulums[0].m2 = ic.m2;
    
    const initialEnergy = simulator.calculateEnergyForPendulum(simulator.pendulums[0]);
    
    // Run simulation
    simulator.play();
    const trajectory = [];
    
    for (let i = 0; i < SAMPLES_PER_TRAJECTORY; i++) {
        simulator.step();
        
        const state = simulator.pendulums[0].state;
        const positions = simulator.getPositionsForPendulum(simulator.pendulums[0]);
        const energy = simulator.calculateEnergyForPendulum(simulator.pendulums[0]);
        
        trajectory.push({
            time: simulator.time,
            theta1: state[0],
            omega1: state[1],
            theta2: state[2],
            omega2: state[3],
            x1: positions.x1,
            y1: positions.y1,
            x2: positions.x2,
            y2: positions.y2,
            energy: energy
        });
    }
    
    const finalEnergy = simulator.calculateEnergyForPendulum(simulator.pendulums[0]);
    const energyDrift = Math.abs((finalEnergy - initialEnergy) / initialEnergy);
    
    // Classify behavior
    const label = classifyBehavior(trajectory, energyDrift);
    const lyapunov = calculateLyapunovApproximation(trajectory);
    
    return {
        trajectoryId,
        initialConditions: ic,
        initialEnergy,
        finalEnergy,
        energyDrift,
        label,
        lyapunov,
        trajectory
    };
}

/**
 * Extract features from trajectory for AI training
 */
function extractFeatures(data) {
    const { initialConditions, trajectory, initialEnergy, lyapunov } = data;
    const ic = initialConditions;
    
    // Initial state features
    const features = [
        ic.theta1, ic.omega1, ic.theta2, ic.omega2,  // Initial state
        ic.l1, ic.l2, ic.m1, ic.m2,                   // Physical parameters
        initialEnergy,                                 // Initial energy
        lyapunov                                       // Chaos measure
    ];
    
    // Statistical features from trajectory
    const theta1_values = trajectory.map(s => s.theta1);
    const theta2_values = trajectory.map(s => s.theta2);
    const omega1_values = trajectory.map(s => s.omega1);
    const omega2_values = trajectory.map(s => s.omega2);
    
    // Mean
    features.push(
        theta1_values.reduce((a,b) => a+b) / theta1_values.length,
        theta2_values.reduce((a,b) => a+b) / theta2_values.length,
        omega1_values.reduce((a,b) => a+b) / omega1_values.length,
        omega2_values.reduce((a,b) => a+b) / omega2_values.length
    );
    
    // Standard deviation
    const mean_theta1 = features[features.length - 4];
    const mean_theta2 = features[features.length - 3];
    const mean_omega1 = features[features.length - 2];
    const mean_omega2 = features[features.length - 1];
    
    features.push(
        Math.sqrt(theta1_values.reduce((a,b) => a + Math.pow(b - mean_theta1, 2), 0) / theta1_values.length),
        Math.sqrt(theta2_values.reduce((a,b) => a + Math.pow(b - mean_theta2, 2), 0) / theta2_values.length),
        Math.sqrt(omega1_values.reduce((a,b) => a + Math.pow(b - mean_omega1, 2), 0) / omega1_values.length),
        Math.sqrt(omega2_values.reduce((a,b) => a + Math.pow(b - mean_omega2, 2), 0) / omega2_values.length)
    );
    
    // Max absolute values
    features.push(
        Math.max(...theta1_values.map(Math.abs)),
        Math.max(...theta2_values.map(Math.abs)),
        Math.max(...omega1_values.map(Math.abs)),
        Math.max(...omega2_values.map(Math.abs))
    );
    
    return features;
}

/**
 * Main data collection function
 */
async function collectData(numTrajectories) {
    console.log(`Starting data collection: ${numTrajectories} trajectories\n`);
    
    const samples = [];
    const startTime = Date.now();
    const labelCounts = { periodic: 0, chaotic: 0, resonant: 0 };
    
    for (let i = 0; i < numTrajectories; i++) {
        if ((i + 1) % 10 === 0) {
            const elapsed = (Date.now() - startTime) / 1000;
            const rate = (i + 1) / elapsed;
            const remaining = (numTrajectories - i - 1) / rate;
            const eta = formatTime(remaining);
            
            process.stdout.write(
                `\rProgress: ${i + 1}/${numTrajectories} (${((i+1)/numTrajectories*100).toFixed(1)}%) | ` +
                `Rate: ${rate.toFixed(1)} traj/s | ETA: ${eta} | ` +
                `P: ${labelCounts.periodic} C: ${labelCounts.chaotic} R: ${labelCounts.resonant}   `
            );
        }
        
        try {
            const ic = generateRandomIC();
            const result = runSimulation(ic, i);
            
            // Create samples from each timestep in the trajectory
            for (let t = 0; t < result.trajectory.length; t++) {
                const point = result.trajectory[t];
                const features = [
                    point.theta1, point.omega1, point.theta2, point.omega2,
                    ic.l1, ic.l2, ic.m1, ic.m2,
                    point.energy,
                    result.lyapunov,
                    point.x1, point.y1, point.x2, point.y2
                ];
                
                samples.push({
                    trajectory_id: i,
                    timestep: t,
                    time: point.time,
                    features: features,
                    label: result.label,
                    metadata: {
                        initialConditions: result.initialConditions,
                        energyDrift: result.energyDrift,
                        lyapunov: result.lyapunov
                    }
                });
            }
            
            labelCounts[result.label]++;
            
        } catch (error) {
            console.error(`\nError in trajectory ${i}:`, error.message);
        }
    }
    
    console.log('\n');
    const totalTime = (Date.now() - startTime) / 1000;
    
    console.log(`\n${'='.repeat(70)}`);
    console.log('Collection Complete!');
    console.log(`Total trajectories: ${numTrajectories}`);
    console.log(`Total samples: ${samples.length}`);
    console.log(`Time elapsed: ${formatTime(totalTime)}`);
    console.log(`Average rate: ${(numTrajectories / totalTime).toFixed(2)} trajectories/s`);
    console.log(`\nLabel distribution:`);
    console.log(`  Periodic: ${labelCounts.periodic} (${(labelCounts.periodic/numTrajectories*100).toFixed(1)}%)`);
    console.log(`  Chaotic: ${labelCounts.chaotic} (${(labelCounts.chaotic/numTrajectories*100).toFixed(1)}%)`);
    console.log(`  Resonant: ${labelCounts.resonant} (${(labelCounts.resonant/numTrajectories*100).toFixed(1)}%)`);
    console.log(`${'='.repeat(70)}\n`);
    
    return samples;
}

/**
 * Save dataset to file
 */
function saveDataset(samples) {
    const timestamp = Date.now();
    const filename = `dataset_${new Date().toISOString().replace(/[:.]/g, '-').split('T')[0]}_${timestamp}.json`;
    const filepath = path.join(DATA_DIR, filename);
    
    const dataset = {
        metadata: {
            generated: new Date().toISOString(),
            simulator: 'double-pendulum',
            version: '1.0',
            totalSamples: samples.length,
            totalTrajectories: Math.max(...samples.map(s => s.trajectory_id)) + 1,
            samplesPerTrajectory: SAMPLES_PER_TRAJECTORY,
            timestep: TIMESTEP,
            featureNames: [
                'theta1', 'omega1', 'theta2', 'omega2',
                'l1', 'l2', 'm1', 'm2',
                'energy', 'lyapunov',
                'x1', 'y1', 'x2', 'y2'
            ],
            labels: ['periodic', 'chaotic', 'resonant']
        },
        samples: samples
    };
    
    console.log(`Saving dataset to: ${filename}`);
    fs.writeFileSync(filepath, JSON.stringify(dataset, null, 2));
    
    const fileSize = fs.statSync(filepath).size;
    console.log(`File size: ${(fileSize / 1024 / 1024).toFixed(2)} MB`);
    console.log(`✓ Dataset saved successfully!\n`);
    
    return filepath;
}

/**
 * Format time in human-readable format
 */
function formatTime(seconds) {
    if (seconds < 60) {
        return `${seconds.toFixed(0)}s`;
    } else if (seconds < 3600) {
        const mins = Math.floor(seconds / 60);
        const secs = Math.floor(seconds % 60);
        return `${mins}m ${secs}s`;
    } else {
        const hours = Math.floor(seconds / 3600);
        const mins = Math.floor((seconds % 3600) / 60);
        return `${hours}h ${mins}m`;
    }
}

/**
 * Main execution
 */
async function main() {
    try {
        // To get 20k samples with 200 samples per trajectory, we need 100 trajectories
        const numTrajectories = Math.ceil(20000 / SAMPLES_PER_TRAJECTORY);
        
        console.log(`Collecting ${numTrajectories} trajectories × ${SAMPLES_PER_TRAJECTORY} timesteps = ~${numTrajectories * SAMPLES_PER_TRAJECTORY} samples\n`);
        
        const samples = await collectData(numTrajectories);
        const filepath = saveDataset(samples);
        
        console.log('='.repeat(70));
        console.log('Data collection completed successfully!');
        console.log(`Dataset saved to: ${filepath}`);
        console.log(`Ready for training with train_double_pendulum.py`);
        console.log('='.repeat(70));
        
    } catch (error) {
        console.error('\nFatal error:', error);
        process.exit(1);
    }
}

// Check if jsdom is installed
try {
    require('jsdom');
} catch (error) {
    console.error('Error: jsdom is not installed.');
    console.error('Please install it with: npm install jsdom');
    process.exit(1);
}

// Run if called directly
if (require.main === module) {
    main();
}

module.exports = { collectData, saveDataset };

File: scripts/python_training/collect_double_pendulum_data.py
#!/./venv/bin/python3
"""
Double Pendulum Data Collection Script (Pure Python)
Generates 20,000 training samples by simulating double pendulum dynamics
"""

import os
import sys
import json
import numpy as np
from datetime import datetime

# Add project root to path
PROJECT_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), '../..'))
sys.path.insert(0, PROJECT_ROOT)

# Data directory
DATA_DIR = os.path.join(PROJECT_ROOT, 'ai_data', 'double-pendulum')
os.makedirs(DATA_DIR, exist_ok=True)

print("=" * 70)
print("Double Pendulum AI Training Data Collection (Python)")
print("=" * 70)
print(f"Target: 20,000 samples")
print(f"Data directory: {DATA_DIR}")
print()


class DoublePendulumSimulator:
    """Pure Python implementation of double pendulum physics"""
    
    def __init__(self, l1=1.0, l2=1.0, m1=1.0, m2=1.0, g=9.81, dt=0.01):
        self.l1 = l1
        self.l2 = l2
        self.m1 = m1
        self.m2 = m2
        self.g = g
        self.dt = dt
        self.time = 0
        self.state = np.array([0.0, 0.0, 0.0, 0.0])  # [theta1, omega1, theta2, omega2]
    
    def derivatives(self, state):
        """Calculate derivatives for double pendulum equations of motion"""
        theta1, omega1, theta2, omega2 = state
        
        delta = theta2 - theta1
        den1 = (self.m1 + self.m2) * self.l1 - self.m2 * self.l1 * np.cos(delta) * np.cos(delta)
        den2 = (self.l2 / self.l1) * den1
        
        dtheta1 = omega1
        dtheta2 = omega2
        
        domega1 = (
            self.m2 * self.l1 * omega1 * omega1 * np.sin(delta) * np.cos(delta) +
            self.m2 * self.g * np.sin(theta2) * np.cos(delta) +
            self.m2 * self.l2 * omega2 * omega2 * np.sin(delta) -
            (self.m1 + self.m2) * self.g * np.sin(theta1)
        ) / den1
        
        domega2 = (
            -self.m2 * self.l2 * omega2 * omega2 * np.sin(delta) * np.cos(delta) +
            (self.m1 + self.m2) * self.g * np.sin(theta1) * np.cos(delta) -
            (self.m1 + self.m2) * self.l1 * omega1 * omega1 * np.sin(delta) -
            (self.m1 + self.m2) * self.g * np.sin(theta2)
        ) / den2
        
        return np.array([dtheta1, domega1, dtheta2, domega2])
    
    def step_rk4(self):
        """Runge-Kutta 4th order integration step"""
        k1 = self.derivatives(self.state)
        k2 = self.derivatives(self.state + self.dt / 2 * k1)
        k3 = self.derivatives(self.state + self.dt / 2 * k2)
        k4 = self.derivatives(self.state + self.dt * k3)
        
        self.state = self.state + (self.dt / 6) * (k1 + 2*k2 + 2*k3 + k4)
        self.time += self.dt
    
    def get_positions(self):
        """Get Cartesian positions of both bobs"""
        theta1, _, theta2, _ = self.state
        
        x1 = self.l1 * np.sin(theta1)
        y1 = -self.l1 * np.cos(theta1)
        
        x2 = x1 + self.l2 * np.sin(theta2)
        y2 = y1 - self.l2 * np.cos(theta2)
        
        return x1, y1, x2, y2
    
    def get_energy(self):
        """Calculate total mechanical energy"""
        theta1, omega1, theta2, omega2 = self.state
        x1, y1, x2, y2 = self.get_positions()
        
        # Kinetic energy
        v1x = self.l1 * omega1 * np.cos(theta1)
        v1y = self.l1 * omega1 * np.sin(theta1)
        v2x = v1x + self.l2 * omega2 * np.cos(theta2)
        v2y = v1y + self.l2 * omega2 * np.sin(theta2)
        
        ke = 0.5 * self.m1 * (v1x**2 + v1y**2) + 0.5 * self.m2 * (v2x**2 + v2y**2)
        
        # Potential energy
        pe = self.m1 * self.g * y1 + self.m2 * self.g * y2
        
        return ke + pe


def generate_random_initial_conditions():
    """Generate random initial conditions for double pendulum"""
    theta1 = (np.random.random() - 0.5) * 2 * np.pi
    theta2 = (np.random.random() - 0.5) * 2 * np.pi
    omega1 = (np.random.random() - 0.5) * 10
    omega2 = (np.random.random() - 0.5) * 10
    
    l1 = 0.8 + np.random.random() * 0.4
    l2 = 0.8 + np.random.random() * 0.4
    m1 = 0.5 + np.random.random() * 1.0
    m2 = 0.5 + np.random.random() * 1.0
    
    return {
        'theta1': theta1,
        'omega1': omega1,
        'theta2': theta2,
        'omega2': omega2,
        'l1': l1,
        'l2': l2,
        'm1': m1,
        'm2': m2
    }


def calculate_lyapunov_approximation(trajectory):
    """Calculate Lyapunov exponent approximation (measure of chaos)"""
    if len(trajectory) < 20:
        return 0.0
    
    sum_log = 0.0
    count = 0
    
    for i in range(10, len(trajectory)):
        d1 = abs(trajectory[i]['theta1'] - trajectory[i-10]['theta1'])
        d2 = abs(trajectory[i]['theta2'] - trajectory[i-10]['theta2'])
        distance = np.sqrt(d1**2 + d2**2)
        
        if distance > 1e-10:
            sum_log += np.log(distance)
            count += 1
    
    return sum_log / count if count > 0 else 0.0


def classify_behavior(trajectory, energy_drift):
    """Classify pendulum behavior based on motion characteristics"""
    if len(trajectory) < 10:
        return 'chaotic'
    
    # Calculate motion statistics
    theta1_changes = [abs(trajectory[i]['theta1'] - trajectory[i-1]['theta1']) 
                     for i in range(1, len(trajectory))]
    theta2_changes = [abs(trajectory[i]['theta2'] - trajectory[i-1]['theta2']) 
                     for i in range(1, len(trajectory))]
    
    avg_theta1_change = np.mean(theta1_changes)
    avg_theta2_change = np.mean(theta2_changes)
    total_theta1_change = sum(theta1_changes)
    total_theta2_change = sum(theta2_changes)
    
    # Calculate standard deviations
    std_theta1 = np.std(theta1_changes)
    std_theta2 = np.std(theta2_changes)
    
    # Check for periodic motion (autocorrelation)
    mid = len(trajectory) // 2
    omega1_values = [t['omega1'] for t in trajectory]
    omega2_values = [t['omega2'] for t in trajectory]
    
    correlation = 0.0
    for i in range(mid):
        diff1 = abs(omega1_values[i] - omega1_values[i + mid])
        diff2 = abs(omega2_values[i] - omega2_values[i + mid])
        correlation += (diff1 + diff2)
    correlation /= mid
    
    # Improved classification logic
    # Periodic: low variability, good autocorrelation
    if correlation < 3.0 and std_theta1 < 0.3 and std_theta2 < 0.3:
        return 'periodic'
    # Resonant: moderate variability, some periodicity
    elif correlation < 5.0 and (std_theta1 < 0.5 or std_theta2 < 0.5):
        return 'resonant'
    # Chaotic: high variability, poor autocorrelation  
    else:
        return 'chaotic'


def simulate_trajectory(ic, num_steps=200):
    """Run a single simulation trajectory"""
    sim = DoublePendulumSimulator(
        l1=ic['l1'], l2=ic['l2'], 
        m1=ic['m1'], m2=ic['m2']
    )
    
    sim.state = np.array([ic['theta1'], ic['omega1'], ic['theta2'], ic['omega2']])
    initial_energy = sim.get_energy()
    
    trajectory = []
    for _ in range(num_steps):
        sim.step_rk4()
        
        x1, y1, x2, y2 = sim.get_positions()
        energy = sim.get_energy()
        
        trajectory.append({
            'time': sim.time,
            'theta1': sim.state[0],
            'omega1': sim.state[1],
            'theta2': sim.state[2],
            'omega2': sim.state[3],
            'x1': x1,
            'y1': y1,
            'x2': x2,
            'y2': y2,
            'energy': energy
        })
    
    final_energy = sim.get_energy()
    energy_drift = abs((final_energy - initial_energy) / initial_energy) if initial_energy != 0 else 0
    
    lyapunov = calculate_lyapunov_approximation(trajectory)
    label = classify_behavior(trajectory, energy_drift)
    
    return {
        'trajectory': trajectory,
        'initial_energy': initial_energy,
        'final_energy': final_energy,
        'energy_drift': energy_drift,
        'lyapunov': lyapunov,
        'label': label
    }


def collect_data(num_trajectories=100, samples_per_trajectory=200):
    """Collect training data"""
    print(f"Collecting {num_trajectories} trajectories × {samples_per_trajectory} timesteps")
    print(f"Total samples: {num_trajectories * samples_per_trajectory}\n")
    
    all_samples = []
    label_counts = {'periodic': 0, 'chaotic': 0, 'resonant': 0}
    
    for traj_id in range(num_trajectories):
        # Print progress
        if (traj_id + 1) % 10 == 0 or traj_id == 0:
            print(f"  Progress: {traj_id + 1}/{num_trajectories} ({(traj_id+1)/num_trajectories*100:.1f}%)")
        
        ic = generate_random_initial_conditions()
        result = simulate_trajectory(ic, samples_per_trajectory)
        
        label_counts[result['label']] += 1
        
        # Create samples from each timestep
        for timestep, point in enumerate(result['trajectory']):
            features = [
                point['theta1'], point['omega1'], point['theta2'], point['omega2'],
                ic['l1'], ic['l2'], ic['m1'], ic['m2'],
                point['energy'],
                result['lyapunov'],
                point['x1'], point['y1'], point['x2'], point['y2']
            ]
            
            all_samples.append({
                'trajectory_id': traj_id,
                'timestep': timestep,
                'time': point['time'],
                'features': features,
                'label': result['label'],
                'metadata': {
                    'initialConditions': ic,
                    'energyDrift': result['energy_drift'],
                    'lyapunov': result['lyapunov']
                }
            })
    
    print(f"\n{'='*70}")
    print("Collection Complete!")
    print(f"Total samples: {len(all_samples)}")
    print(f"\nLabel distribution:")
    for label, count in label_counts.items():
        percent = (count / num_trajectories) * 100
        print(f"  {label.capitalize()}: {count} ({percent:.1f}%)")
    print('='*70)
    
    return all_samples, label_counts


def save_dataset(samples):
    """Save dataset to JSON file"""
    timestamp = int(datetime.now().timestamp())
    date_str = datetime.now().strftime('%Y-%m-%d')
    filename = f"dataset_{date_str}_{timestamp}.json"
    filepath = os.path.join(DATA_DIR, filename)
    
    dataset = {
        'metadata': {
            'generated': datetime.now().isoformat(),
            'simulator': 'double-pendulum',
            'version': '1.0',
            'totalSamples': len(samples),
            'totalTrajectories': max(s['trajectory_id'] for s in samples) + 1,
            'samplesPerTrajectory': 200,
            'timestep': 0.01,
            'featureNames': [
                'theta1', 'omega1', 'theta2', 'omega2',
                'l1', 'l2', 'm1', 'm2',
                'energy', 'lyapunov',
                'x1', 'y1', 'x2', 'y2'
            ],
            'labels': ['periodic', 'chaotic', 'resonant']
        },
        'samples': samples
    }
    
    print(f"\nSaving dataset to: {filename}")
    with open(filepath, 'w') as f:
        json.dump(dataset, f, indent=2)
    
    file_size = os.path.getsize(filepath)
    print(f"File size: {file_size / 1024 / 1024:.2f} MB")
    print(f"✓ Dataset saved successfully!")
    
    return filepath


def main():
    """Main execution"""
    try:
        # Collect 100 trajectories × 200 timesteps = 20,000 samples
        samples, label_counts = collect_data(num_trajectories=100, samples_per_trajectory=200)
        
        # Save dataset
        filepath = save_dataset(samples)
        
        print("\n" + "="*70)
        print("Data collection completed successfully!")
        print(f"Dataset: {filepath}")
        print("Ready for training with train_double_pendulum.py")
        print("="*70)
        
    except KeyboardInterrupt:
        print("\n\n⚠ Collection interrupted by user")
        sys.exit(0)
    except Exception as e:
        print(f"\nERROR: {str(e)}")
        import traceback
        traceback.print_exc()
        sys.exit(1)


if __name__ == '__main__':
    main()
